### PR TITLE
BUG: OBJECT_nonzero does not check for exceptions

### LINKED ISSUE #9295: `SystemError` raised by `nonzero([np.array([1, 2]), 3])`
Here's the full traceback in an ipython session (Python 3.5.2):

```
In [29]: np.__version__
Out[29]: '1.14.0.dev0+518e074'

In [30]: np.nonzero([np.array([1, 2]), 3])
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
/Users/warren/miniconda3numpy/lib/python3.5/site-packages/numpy-1.14.0.dev0+518e074-py3.5-macosx-10.6-x86_64.egg/numpy/core/fromnumeric.py in _wrapfunc(obj, method, *args, **kwds)
     56     try:
---> 57         return getattr(obj, method)(*args, **kwds)
     58 

AttributeError: 'list' object has no attribute 'nonzero'

During handling of the above exception, another exception occurred:

ValueError                                Traceback (most recent call last)
ValueError: The truth value of an array with more than one element is ambiguous. Use a.any() or a.all()

During handling of the above exception, another exception occurred:

SystemError                               Traceback (most recent call last)
<ipython-input-30-f6cfb4ce337f> in <module>()
----> 1 np.nonzero([np.array([1, 2]), 3])

/Users/warren/miniconda3numpy/lib/python3.5/site-packages/numpy-1.14.0.dev0+518e074-py3.5-macosx-10.6-x86_64.egg/numpy/core/fromnumeric.py in nonzero(a)
   1560 
   1561     """
-> 1562     return _wrapfunc(a, 'nonzero')
   1563 
   1564 

/Users/warren/miniconda3numpy/lib/python3.5/site-packages/numpy-1.14.0.dev0+518e074-py3.5-macosx-10.6-x86_64.egg/numpy/core/fromnumeric.py in _wrapfunc(obj, method, *args, **kwds)
     65     # a downstream library like 'pandas'.
     66     except (AttributeError, TypeError):
---> 67         return _wrapit(obj, method, *args, **kwds)
     68 
     69 

/Users/warren/miniconda3numpy/lib/python3.5/site-packages/numpy-1.14.0.dev0+518e074-py3.5-macosx-10.6-x86_64.egg/numpy/core/fromnumeric.py in _wrapit(obj, method, *args, **kwds)
     45     except AttributeError:
     46         wrap = None
---> 47     result = getattr(asarray(obj), method)(*args, **kwds)
     48     if wrap:
     49         if not isinstance(result, mu.ndarray):

SystemError: <built-in method nonzero of numpy.ndarray object at 0x105c963a0> returned a result with an error set
```

The problem was originally noticed in a stackoverflow question:  https://stackoverflow.com/questions/44477539/an-error-in-matplotlib-related-to-numpy-roots

### PR DESCRIPTION
**Pull request by ahaldane**
Fixes #9295

### CODE REVIEW

**Comment by eric-wieser** [comment]
Shouldn't this exception be raised somehow, rather than silenced? Perhaps the caller should be expected to inspect `PyError_Occurred`

File: numpy/core/tests/test_numeric.py [hunk 1]
```diff
@@ -1107,6 +1107,11 @@ def test_array_method(self):
 
         assert_equal(m.nonzero(), tgt)
 
+    def test_nonzero_invalid_object(self):
+        # gh-9295
+        a = np.array([np.array([1, 2]), 3])
+        assert_equal(a.nonzero(), (array([0, 1]),))
+
```
Comment (eric-wieser) [inline numpy/core/tests/test_numeric.py]: Missing `np.`

**Comment by ahaldane** [comment]
It seemed like a much more invasive change to error check in all the callers, which motivated me to do it this way.

Also, we are looking for nonzero elements of the array, so it seems fair to interpret elements which cannot be evaluated as True (because of some failure) as non-zero.

File: numpy/core/tests/test_numeric.py (hunk 1 above)
Comment (ahaldane) [inline numpy/core/tests/test_numeric.py]: thanks

**Comment by ahaldane** [comment]
Hmm actually maybe you are right, based on the failing test...

**Comment by ahaldane** [comment]
OK, so looking through all the files that call nonzero, most of them actually *did* check for errors. 

There was only one place where the error wasn't checked.

I checked that all the files `item_selection.c`, `scalarmath.c.src`,  `arraytypes.c.src`, `methods.c`, `sequence.c`, and `number.c` either explicitly account for errors or pass on the -1 value. In `item_selection.c:PyArray_Nonzero` I assume that if `PyArray_CountNonzero` did not fail, then subsequent calls to `nonzero` on the same array cannot fail.

File: numpy/core/src/multiarray/item_selection.c [hunk 2]
```diff
@@ -2099,7 +2099,7 @@ PyArray_CountNonzero(PyArrayObject *self)
             data += stride;
         }
 
-        return nonzero_count;
+        return PyErr_Occurred() ? -1 : nonzero_count;
```
Comment (eric-wieser) [inline numpy/core/src/multiarray/item_selection.c]: Should this be inside the `while` loop?

**Comment by ahaldane** [comment]
Oh... it won't let me reopen on a force-pushed branch?

I'm going to have top open a new issue.

**Pull request closed without merging**